plugins/state_history_plugin/include/eosio/state_history_plugin/session.hpp

```diff
@@ -16,691 +19,327 @@ extern const char* const state_history_plugin_abi;
+                           self.current_blocks_request.start_block_num = std::min(self.current_blocks_request.start_block_num, haveit.block_num);
+                     }
+                     self.current_blocks_request.have_positions.clear();
+                  },
+                  [&self](const get_blocks_ack_request_v0& gbar0) {
+                     self.send_credits += gbar0.num_messages;
+                  }
+               }, req);
+               co_return;
+            }, boost::asio::use_awaitable);
+
+            awake_if_idle();
          }
-      }
+      });
+   }
 
-      if (block_id) {
-         result.this_block  = state_history::block_position{to_send_block_num, *block_id};
-         auto prev_block_id = plugin.get_block_id(to_send_block_num - 1);
-         if (prev_block_id)
-            result.prev_block = state_history::block_position{to_send_block_num - 1, *prev_block_id};
-         if (request.fetch_block) {
-            uint32_t block_num = block ? block->block_num() : 0; // block can be nullptr in testing
-            plugin.get_block(to_send_block_num, block_num, block, result.block);
-         }
-         if (request.fetch_traces && plugin.get_trace_log())
-            result.traces.emplace();
-         if (request.fetch_deltas && plugin.get_chain_state_log())
-            result.deltas.emplace();
-         if constexpr (std::is_same_v<T, state_history::get_blocks_result_v1>) {
-            if (fetch_finality_data && plugin.get_finality_data_log()) {
-               result.finality_data.emplace();
-            }
-         }
-      }
-      ++to_send_block_num;
-
-      // during syncing if block is older than 5 min, log every 1000th block
-      bool fresh_block = fc::time_point::now() - plugin.get_head_block_timestamp() < fc::minutes(5);
-      if (fresh_block || (result.this_block && result.this_block->block_num % 1000 == 0)) {
-         fc_ilog(plugin.get_logger(),
-                 "pushing result "
-                 "{\"head\":{\"block_num\":${head}},\"last_irreversible\":{\"block_num\":${last_irr}},\"this_block\":{"
-                 "\"block_num\":${this_block}, \"block_id\":${this_id}}} to send queue",
-                 ("head", result.head.block_num)("last_irr", result.last_irreversible.block_num)
-                 ("this_block", result.this_block ? result.this_block->block_num : fc::variant())
-                 ("this_id", result.this_block ? fc::variant{result.this_block->block_id} : fc::variant{}));
-      }
+   get_status_result_v1 fill_current_status_result() {
+      get_status_result_v1 ret;
 
-      --request.max_messages_in_flight;
-      need_to_send_update = to_send_block_num <= current &&
-                            to_send_block_num < request.end_block_num;
+      ret.head              = {controller.head_block_num(), controller.head_block_id()};
+      ret.last_irreversible = {controller.last_irreversible_block_num(), controller.last_irreversible_block_id()};
```

Comment:
This actually looks like an error -- `fill_current_status_result()` should be called which still in the main thread. It'll also touch things like `trace_log` which are documented as being main thread only

Reply:
I moved `fill_current_status_result()` to being in main thread on 637fff2

I didn't change the pattern to `num_from_id()` as you suggested because I like the current aggregate initialization and I'm not sure how to do a single `controller` call and maintain that aggregate initialization structure at first glance

Reply:
Not needed if called on the main thread as the values can't change. It is a bit more expensive to call them separately (two mutex locks), but not a huge deal.